python/lsst/daf/butler/registries/sqlRegistry.py

```diff
@@ -93,6 +93,118 @@ def transaction(self):
+            all fields to be held in the database.
+        key : `str`
+            The name of the field to be used as the dictionary key.  Must not
+            be present in ``value._fields``.
+        value : `type`
+            The type used for the dictionary's values, typically a
+            `~collections.namedtuple`.  Must have a ``_fields`` class
+            attribute that is a tuple of field names (i.e. as defined by
+            `~collections.namedtuple`); these field names must also appear
+            in the ``types`` arg, and a `_make` attribute to construct it
+            from a sequence of values (again, as defined by
+            `~collections.namedtuple`).
+        """
+        # We need to construct a temporary config for the table value because
+        # SqlRegistryDatabaseDict.__init__ is required to take a config so it
+        # can be called by DatabaseDict.fromConfig as well.
+        # I suppose we could have Registry.makeDatabaseDict take a config as
+        # well, since it"ll also usually be called by DatabaseDict.fromConfig,
+        # but I strongly believe in having signatures that only take what they
+        # really need.
+        config = Config()
+        config["table"] = table
+        return SqlRegistryDatabaseDict(config, types=types, key=key, value=value, registry=self)
+
+    def find(self, collection, datasetType, dataId):
+        """Lookup a dataset.
+
+        This can be used to obtain a `DatasetRef` that permits the dataset to
+        be read from a `Datastore`.
+
+        Parameters
+        ----------
+        collection : `str`
+            Identifies the Collection to search.
+        datasetType : `DatasetType`
+            The `DatasetType`.
+        dataId : `dict`
+            A `dict` of `DataUnit` link name, value pairs that label the
+            `DatasetRef` within a Collection.
+
+        Returns
+        -------
+        ref : `DatasetRef`
+            A ref to the Dataset, or `None` if no matching Dataset
+            was found.
+
+        Raises
+        ------
+        ValueError
+            If dataId is invalid.
+        """
+        self._validateDataId(datasetType, dataId)
+        datasetTable = self._schema.tables["Dataset"]
+        datasetCollectionTable = self._schema.tables["DatasetCollection"]
+        dataIdExpression = and_((self._schema.dataUnits.links[name] == dataId[name]
+                                 for name in self._schema.dataUnits.getPrimaryKeyNames(
+                                     datasetType.dataUnits)))
+        result = self._connection.execute(select([datasetTable.c.dataset_id]).select_from(
+            datasetTable.join(datasetCollectionTable)).where(and_(
+                datasetTable.c.dataset_type_name == datasetType.name,
+                datasetCollectionTable.c.collection == collection,
+                dataIdExpression))).fetchone()
+        # TODO update unit values and add Run, Quantum and assembler?
```

Comment:
We should probably have some options in the public API controlling how much we do here.  Let's try just having a `full=False` option, where calling with `full=True` would add optional DataUnit link values (e.g. `physical_filter` when `visit` is present), Quantums, and (until we get rid of it) assembler.

Reply:
Agreed, but probably best done on a new ticket.